File: src/SMS.Modules.Finance/Services/MasterProductLedgerService.cs
using System.Data.Common;
using Microsoft.EntityFrameworkCore;
using SMS.Modules.Finance.Data;
using SMS.Modules.Finance.Domain;
using SMS.Modules.Finance.Models;
using SMS.Shared.Common;
using SMS.Shared.Pagination;

namespace SMS.Modules.Finance.Services;

// FSD Addendum 24 (ML-003/ML-004) — write side implements the SMS.Shared.Common contract so
// SMS.Modules.Inventory (and, transitively, Warehouse/Material) can post product movements without
// referencing Finance directly (internal — only the shared interface is public). Read side
// (IMasterProductLedgerQueryService) is public since it only deals in public model types and is
// consumed directly by MasterProductLedgerController, which lives in this same module.
internal sealed class MasterProductLedgerService : IMasterProductLedgerService, IMasterProductLedgerQueryService
{
    private readonly FinanceDbContext _db;

    public MasterProductLedgerService(FinanceDbContext db) => _db = db;

    public async Task PostMovementAsync(ProductMovementContext context, DbTransaction? transaction = null)
    {
        var qty = context.QuantityIn ?? context.QuantityOut ?? 0m;

        var entry = new MasterProductLedger
        {
            LedgerId        = Guid.NewGuid(),
            ProductId       = context.ProductId,
            ProductCode     = context.ProductCode,
            ProductName     = context.ProductName,
            CategoryId      = context.CategoryId,
            CategoryName    = context.CategoryName,
            WarehouseId     = context.WarehouseId,
            WarehouseName   = context.WarehouseName,
            TransactionDate = DateTime.UtcNow,
            TransactionType = context.TransactionType,
            ReferenceType   = context.ReferenceType,
            ReferenceId     = context.ReferenceId,
            ReferenceNumber = context.ReferenceNumber,
            QuantityIn      = context.QuantityIn,
            QuantityOut     = context.QuantityOut,
            UnitCost        = context.UnitCost,
            TotalValue      = qty * context.UnitCost,
            SourceType      = context.SourceType,
            SourceName      = context.SourceName,
            DestinationType = context.DestinationType,
            DestinationName = context.DestinationName,
            Notes           = context.Notes,
            CreatedBy       = context.CreatedBy,
            CreatedDate     = DateTime.UtcNow
        };

        _db.MasterProductLedgers.Add(entry);

        if (transaction is not null)
        {
            // Enlist THIS DbContext (a different type/schema than the caller's) into the caller's
            // already-open ambient transaction, so this insert commits — or rolls back — atomically
            // with the InventoryLedgerEntry and whatever business write triggered it. Same technique
            // already used across DbContexts in MaterialReturnService/MivService.PostAsync.
            // contextOwnsConnection: false — this context must never close a connection it doesn't own.
            _db.Database.SetDbConnection(transaction.Connection!, contextOwnsConnection: false);
            await _db.Database.UseTransactionAsync(transaction);
        }

        await _db.SaveChangesAsync();
    }

    // ── ML-004 read side ─────────────────────────────────────────────────────

    private IQueryable<MasterProductLedger> ApplyFilter(MasterProductLedgerFilter filter)
    {
        var q = _db.MasterProductLedgers.AsQueryable();

        if (filter.DateFrom.HasValue)    q = q.Where(e => e.TransactionDate >= filter.DateFrom.Value);
        if (filter.DateTo.HasValue)      q = q.Where(e => e.TransactionDate <= filter.DateTo.Value);
        if (filter.ProductId.HasValue)   q = q.Where(e => e.ProductId == filter.ProductId.Value);
        if (filter.CategoryId.HasValue)  q = q.Where(e => e.CategoryId == filter.CategoryId.Value);
        if (filter.WarehouseId.HasValue) q = q.Where(e => e.WarehouseId == filter.WarehouseId.Value);
        if (!string.IsNullOrWhiteSpace(filter.TransactionType)) q = q.Where(e => e.TransactionType == filter.TransactionType);
        if (!string.IsNullOrWhiteSpace(filter.SourceType))      q = q.Where(e => e.SourceType == filter.SourceType);
        if (!string.IsNullOrWhiteSpace(filter.DestinationType)) q = q.Where(e => e.DestinationType == filter.DestinationType);

        return q;
    }

    public async Task<PaginatedResponse<MasterProductLedgerEntryModel>> GetProductLedgerAsync(MasterProductLedgerFilter filter)
    {
        var query = ApplyFilter(filter);

        var total    = await query.CountAsync();
        var page     = Math.Max(1, filter.Page);
        var pageSize = Math.Clamp(filter.PageSize, 1, 200);

        var entities = await query
            .OrderByDescending(e => e.TransactionDate)
            .ThenByDescending(e => e.Id)
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .ToListAsync();

        return new PaginatedResponse<MasterProductLedgerEntryModel>
        {
            Data         = entities.Select(ToModel).ToList(),
            TotalRecords = total,
            Page         = page,
            PageSize     = pageSize,
            TotalPages   = (int)Math.Ceiling(total / (double)pageSize)
        };
    }

    public async Task<MasterProductLedgerSummaryModel> GetProductLedgerSummaryAsync(MasterProductLedgerFilter filter)
    {
        var filtered = ApplyFilter(filter);

        var totalReceipts = await filtered.SumAsync(e => (decimal?)(e.QuantityIn ?? 0m)) ?? 0m;
        var totalIssues   = await filtered.SumAsync(e => (decimal?)(e.QuantityOut ?? 0m)) ?? 0m;
        var totalValue    = await filtered.SumAsync(e => (decimal?)e.TotalValue) ?? 0m;

        return new MasterProductLedgerSummaryModel
        {
            TotalReceiptsQty = totalReceipts,
            TotalIssuesQty   = totalIssues,
            NetMovement      = totalReceipts - totalIssues,
            TotalValueMoved  = totalValue
        };
    }

    public async Task<List<MasterProductLedgerEntryModel>> GetProductJourneyAsync(int productId)
    {
        var entities = await _db.MasterProductLedgers
            .Where(e => e.ProductId == productId)
            .OrderBy(e => e.TransactionDate)
            .ThenBy(e => e.Id)
            .ToListAsync();

        return entities.Select(ToModel).ToList();
    }

    private static MasterProductLedgerEntryModel ToModel(MasterProductLedger e) => new()
    {
        LedgerId        = e.LedgerId,
        ProductId       = e.ProductId,
        ProductCode     = e.ProductCode,
        ProductName     = e.ProductName,
        CategoryId      = e.CategoryId,
        CategoryName    = e.CategoryName,
        WarehouseId     = e.WarehouseId,
        WarehouseName   = e.WarehouseName,
        TransactionDate = e.TransactionDate,
        TransactionType = e.TransactionType,
        ReferenceType   = e.ReferenceType,
        ReferenceId     = e.ReferenceId,
        ReferenceNumber = e.ReferenceNumber,
        QuantityIn      = e.QuantityIn,
        QuantityOut     = e.QuantityOut,
        UnitCost        = e.UnitCost,
        TotalValue      = e.TotalValue,
        SourceType      = e.SourceType,
        SourceName      = e.SourceName,
        DestinationType = e.DestinationType,
        DestinationName = e.DestinationName,
        Notes           = e.Notes,
        CreatedBy       = e.CreatedBy,
        CreatedDate     = e.CreatedDate
    };
}

File: SupplyChainFrontend/src/app/services/master-product-ledger.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ApiResponse, PaginatedResponse } from './demand.service';

const BASE = 'https://localhost:52800/api/inventory/master-product-ledger';

export interface MasterProductLedgerEntryModel {
  ledgerId: string;
  productId: number;
  productCode: string;
  productName: string;
  categoryId?: number;
  categoryName?: string;
  warehouseId: number;
  warehouseName: string;
  transactionDate: string;
  transactionType: string;
  referenceType: string;
  referenceId: string;
  referenceNumber: string;
  quantityIn?: number;
  quantityOut?: number;
  unitCost: number;
  totalValue: number;
  sourceType: string;
  sourceName?: string;
  destinationType: string;
  destinationName?: string;
  notes?: string;
  createdBy: number;
  createdDate: string;
}

export interface MasterProductLedgerFilter {
  dateFrom?: string;
  dateTo?: string;
  productId?: number;
  categoryId?: number;
  warehouseId?: number;
  transactionType?: string;
  sourceType?: string;
  destinationType?: string;
  page?: number;
  pageSize?: number;
}

export interface MasterProductLedgerSummaryModel {
  totalReceiptsQty: number;
  totalIssuesQty: number;
  netMovement: number;
  totalValueMoved: number;
}

@Injectable({ providedIn: 'root' })
export class MasterProductLedgerService {
  constructor(private http: HttpClient) {}

  private buildParams(filter: MasterProductLedgerFilter): HttpParams {
    let params = new HttpParams();
    if (filter.dateFrom)        params = params.set('dateFrom', filter.dateFrom);
    if (filter.dateTo)          params = params.set('dateTo', filter.dateTo);
    if (filter.productId != null)   params = params.set('productId', filter.productId);
    if (filter.categoryId != null)  params = params.set('categoryId', filter.categoryId);
    if (filter.warehouseId != null) params = params.set('warehouseId', filter.warehouseId);
    if (filter.transactionType) params = params.set('transactionType', filter.transactionType);
    if (filter.sourceType)      params = params.set('sourceType', filter.sourceType);
    if (filter.destinationType) params = params.set('destinationType', filter.destinationType);
    if (filter.page)     params = params.set('page', filter.page);
    if (filter.pageSize) params = params.set('pageSize', filter.pageSize);
    return params;
  }

  getLedger(filter: MasterProductLedgerFilter): Observable<ApiResponse<PaginatedResponse<MasterProductLedgerEntryModel>>> {
    return this.http.get<ApiResponse<PaginatedResponse<MasterProductLedgerEntryModel>>>(BASE, { params: this.buildParams(filter) });
  }

  getSummary(filter: MasterProductLedgerFilter): Observable<ApiResponse<MasterProductLedgerSummaryModel>> {
    return this.http.get<ApiResponse<MasterProductLedgerSummaryModel>>(`${BASE}/summary`, { params: this.buildParams(filter) });
  }

  getProductJourney(productId: number): Observable<ApiResponse<MasterProductLedgerEntryModel[]>> {
    return this.http.get<ApiResponse<MasterProductLedgerEntryModel[]>>(`${BASE}/product-journey/${productId}`);
  }

  exportPdf(filter: MasterProductLedgerFilter): Observable<Blob> {
    return this.http.get(`${BASE}/export/pdf`, { params: this.buildParams(filter), responseType: 'blob' });
  }

  exportExcel(filter: MasterProductLedgerFilter): Observable<Blob> {
    return this.http.get(`${BASE}/export/excel`, { params: this.buildParams(filter), responseType: 'blob' });
  }
}
